Add Base.allunique

tpapp opened the pull request:
Allows testing if elements in a collection are distinct (when compared
with isequal). Terminates early in case of repeated elements, has
methods for collections which are by construction distinct. Also added
tests.

See discussion at issue #15803, which this PR fixes.

tpapp committed:
Add Base.alldistinct.

Allows testing if elements in a collection are distict (when compared
with isequal). Terminates early in case of repeated elements, has
methods for collections which are by construction distinct. Also added
tests.

See discussion at issue #15803.

tpapp committed:
Corrected typo in function name.

The typo also hid the error in the function body that was not caught by
the unit tests, now that is also fixed.

hayd commented on base/set.jl:
```diff
@@ -141,6 +141,27 @@ function unique(f::Callable, C)
+
+alldistinct(::Set) = true
+
+alldistinct(r::Range) = (length(r) == 1) || (step(r) > 0)
```
This should probably be:

    (step(r) != 0) || (length(r) <= 1)

or potentially:

    alldistinct{T}(r::Range{T}) = (step(r) !== zero(T)) || (length(r) <= one(T))

and add tests for negative step and zero length ranges.

tpapp replied:
Thanks. Made the change, with tests.

tpapp committed:
Make alldistinct work for negative and 0-length ranges.

Suggestion by @hayd. Add tests for negative ranges, and also empty
collections.

tkelman commented on base/set.jl:
```diff
@@ -141,6 +141,27 @@ function unique(f::Callable, C)
 end
 
+"""
+    alldistinct(itr)
```
needs to be added to rst docs too

tpapp replied:
Done, thanks for the reminder.

tpapp committed:
Added alldistinct to rst source.

tpapp commented:
Is there anything preventing a merge? I understand the core devs are busy so it might take time, but in case I missed something that prevents a merge that I can fix please let me know.

tkelman commented:
I think this is good. If no one has any more feedback, I'll merge soon.

Note for next time, it can help in keeping things organized if you use a new branch (other than master of your fork) for making each pull request.

tkelman commented on base/set.jl:
```diff
@@ -141,6 +141,27 @@ function unique(f::Callable, C)
+"""
+    alldistinct(itr)
+
+Test if all values from `itr` are distinct, when compared with `isequal`.
```
hm, spoke too soon maybe - technically this isn't testing using `isequal` until #9381 gets resolved, right?

nalimilan replied:
`?in` says that for sets, `isequal is used`, so that looks OK.

stevengj replied:
s/Test if/Return `true` if/

No comma before "when".

tpapp replied:
@stevengj: thanks, pushed minor change correcting that.
@tkelman: Sorry that I did not branch, I am still learning the ropes, will do it next time.

I hope that this can be merged now, but let me know if I should do anything else.

tpapp committed:
Minor correction to docstring.

tpapp committed:
Replaced `alldistinct` with `allunique`.

Following discussion for issue #15803, this emerged as the preferred alternative.

nalimilan commented:
Thanks. Can you squash the commits?

KristofferC commented:
Can't we start using githubs own squasher?

stevengj commented:
Yes, PRs no longer have to squash commits manually, thank goodness.

StefanKarpinski commented:
AppVeyor seems to be stuck, but it got to SUCCESS, so I'm calling it a pass.

tkelman commented:
when did this get renamed to allunique? I thought alldistinct was clearer

StefanKarpinski commented:
There was a whole conversation about it.

okvs commented:
There's a problem here which also popped up in #15009 , namely that you can have 'oversampled' ranges of floats:
```julia
r = linspace(1.0, nextfloat(1.0), 10)
>> allunique(r)
true
>> allunique(collect(r))
false
```
As suggested in https://github.com/JuliaLang/julia/pull/15009#discussion_r55024453  you can prevent this with the following check:
```julia
last(r) >= nextfloat(first(r), length(r) - 1)
```

One more thing: This comparison
```julia
length(r) <= one(T)
````
seems strange to me. `length` always returns an `Int`, right? So it should be `length(r) <= 1`.

(I tried all this on 0.4.5, so apologies if `linspace` or `length` have somehow changed in 0.5.)
